### DISCUSSION TITLE
Only allow certain responses

### DISCUSSION THREAD

**Discussion by mrhunsaker** [zauberzeug/nicegui#902] [Q&A] [answered]
Is there a way to set ui.number to only allow numbers (integers) 0, 1, 2, and 3? 
I can see how validation could work by throwing up a warning, but it is much easier if I write it so that no other inputs would be allowed, as that would keep people I work with from typing in invalid responses. (0-3 are values for a scoring RUBRIC in an education context)

**Comment by CatamountJack** [comment, marked as answer]
If you have a simple range, `ui.number(min=0, max=3, format="%1.0f")` may do what you need.  If someone types a decimal, it will convert it to given 'format', but it may round (2.7 --> 3).

For a non-continuous range or to avoid the rounding issue, there is likely a simpler solution, but here's an option:

```python
class Limiter:

    allowable_values = (None, 0, 1, 2, 3)
    last_valid = None

    @classmethod
    def update(cls, new_val):
        cls.last_valid = new_val

    def __init__(self, check_val):
        if check_val in self.allowable_values:
            self.update(int(check_val))


def constrain():
    score.value = Limiter(score.value).last_valid


score = ui.number(format="%1.0f").on("update:model-value", constrain)

ui.run()
```
